Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift

```diff
@@ -185,6 +185,36 @@ extension BrowserViewController: WKNavigationDelegate {
     let isPrivateBrowsing = PrivateBrowsingManager.shared.isPrivateBrowsing
     let tab = tabManager[webView]
 
+    let domainForRequestURL = Domain.getOrCreate(
+      forUrl: url,
+      persistent: !isPrivateBrowsing
+    )
+
+    // Debouncing logic
+    // Handle debouncing for main frame only and only if the site (etld+1) changes
+    // We also only handle `http` and `https` requests
+    if url.isWebPage(includeDataURIs: false),
+       let currentURL = tab?.webView?.url, currentURL.baseDomain != url.baseDomain,
+       domainForRequestURL.isShieldExpected(.AdblockAndTp, considerAllShieldsOption: true),
+       navigationAction.targetFrame?.isMainFrame == true,
+       let redirectURL = DebouncingResourceDownloader.shared.redirectURL(for: url) {
+      // Cancel the original request. We don't want it to load as it's tracking us
+      decisionHandler(.cancel, preferences)
+
+      // We only include trusted headers on cross origin requests
+      // TODO: @JS Look at possibly getting a list of trusted headers from brave-core
+      // For now we only allow the `Referrer`. The browser will add the rest.
+      var modifiedRequest = URLRequest(url: redirectURL)
+      let trustedHeaderKeys = Set(["Referer"])
```

Comment:
if the set is always of size one, couldn't this (and the below loop) be rewritten to just check for `Refer` and simplify all this?